Core/include/Acts/Navigation/MultiNavigationPolicy.hpp

```diff
@@ -0,0 +1,77 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Navigation/INavigationPolicy.hpp"
+
+namespace Acts {
+
+/// Base class for multi navigation policies
+class MultiNavigationPolicyBase : public INavigationPolicy {};
+
+/// Combined navigation policy that calls all contained other navigation
+/// policies. This class only works with policies complying with
+/// `NavigationPolicyConcept`, which means that they have a conventional
+/// `updateState` method.
+///
+/// Internally, this uses a delegate chain factory to produce an unrolled
+/// delegate chain.
+///
+/// @tparam Policies The navigation policies to be combined
+template <NavigationPolicyConcept... Policies>
+  requires(sizeof...(Policies) > 0)
+class MultiNavigationPolicy final : public MultiNavigationPolicyBase {
```

Comment:
Why is this chaining necessary again?

Reply:
To be able to combine e.g. a portal lookup with an unrelated sensor lookup. 

The alternative is writing a mega policy that can do both, or use separate user defined composition or class inheritance. 

The idea with the chain here is to provide a pattern to support this, rather than having to implement this on for each policy you might want to combine. 

In practice, I imagine using the try all portal policy in combination with a layer like policy most of the time, but I'd like to avoid the latter having to also implement portal lookup.

Reply:
I just see the use case rather narrow and the interface and code burden rather big. Instead of chaining one could have an additional navigation policy in the surface array lookup for example to delegate what to do about the portals.

My main concern is that we shape the interface of the initialize candidates call so it can be chained which results in a gen2 like navigation delegate which can practically access and mutate anything.

Reply:
> Instead of chaining one could have an additional navigation policy in the surface array lookup for example to delegate what to do about the portals.

But that requires each policy to foresee remaining strategies, or implement another type erasure mechanism, i.e. by using another delegate or a nested class. 

I think having this as a system, rather than leaving it up to each policy would allow us to keep the policies simpler by making them do only one thing.

Reply:
Also let me clarify that the chaining imo does not contradict the goal of a well defined interface I think.

The only complexity it adds imo is that indeed returning by value is tricky. 

With narrow purpose function calls, we can design the interface to be clear on how the multiple results can be accumulated. Like an output vector, that the policies push back into. 

I agree that we want to avoid just passing through complex state and allow each policy to mutate arbitrarily.

Reply:
I think my main problem with the chaining is that I only see a single use case and it looks like we overgeneralize this similar to what is done in the propagation. We allow for N arbitrary actors but in reality we only ever use one because they would only conflict with each other and it is just not useful to have more than one.

I think we only ever 1. provide surface candidates and/or 2. provide portal candidates and we have a common way of doing this individually so we want to compose that.

Reply:
Right but then what if there are many portals? Try all? What if I have a layer binning and many portals? Do I now implement four variations:

- layer lookup + smart Portal lookup
- layer lookup + try all portals
- try all sensors + smart Portal lookup
- try all sensors + try all portals

What if I add another policy as an experiment? Do I get to reimplement the other stuff as well? Sure, we can architect each strategy as a thing the combined policy call into, but that's exactly what the chaining does, in a structured and systematic way.

Reply:
> I think we only ever 1. provide surface candidates and/or 2. provide portal candidates and we have a common way of doing this individually so we want to compose that.

What I'm proposing here **is that composition** in a structured way. But you don't have to invent your own interface every time, or make lots duplicates or lists of possible configurations, you just plug it in the chain and that's it.